x-pack/legacy/plugins/infra/public/components/inventory/toolbars/container_toolbar.tsx

```diff
@@ -0,0 +1,113 @@
+import { InfraSnapshotMetricType } from '../../../graphql/types';
+import { fieldToName } from '../../waffle/lib/field_to_display_name';
+
+export const ContainerToolbar = (props: ToolbarProps) => {
+  const options = useMemo(() => {
+    const CPUUsage = i18n.translate('xpack.infra.waffle.metricOptions.cpuUsageText', {
+      defaultMessage: 'CPU usage',
+    });
+
+    const MemoryUsage = i18n.translate('xpack.infra.waffle.metricOptions.memoryUsageText', {
+      defaultMessage: 'Memory usage',
+    });
+
+    const InboundTraffic = i18n.translate('xpack.infra.waffle.metricOptions.inboundTrafficText', {
+      defaultMessage: 'Inbound traffic',
+    });
+
+    const OutboundTraffic = i18n.translate('xpack.infra.waffle.metricOptions.outboundTrafficText', {
+      defaultMessage: 'Outbound traffic',
+    });
+
+    return [
+      {
+        text: CPUUsage,
+        value: InfraSnapshotMetricType.cpu,
+      },
+      {
+        text: MemoryUsage,
+        value: InfraSnapshotMetricType.memory,
+      },
+      {
+        text: InboundTraffic,
+        value: InfraSnapshotMetricType.rx,
+      },
+      {
+        text: OutboundTraffic,
+        value: InfraSnapshotMetricType.tx,
+      },
+    ];
+  }, []);
+
+  const groupByOptions = useMemo(() => {
+    const mapFieldToOption = (field: string) => ({
+      text: fieldToName(field),
+      field,
+    });
+    return [
+      'host.name',
+      'cloud.availability_zone',
+      'cloud.machine.type',
+      'cloud.project.id',
+      'cloud.provider',
+      'service.type',
+    ].map(mapFieldToOption);
+  }, []);
+  return (
+    <Toolbar>
+      <EuiFlexGroup alignItems="center" gutterSize="m">
+        <WithSource>
+          {({ createDerivedIndexPattern }) => (
+            <WithWaffleOptions>
+              {({
+                changeMetric,
+                changeGroupBy,
+                changeCustomOptions,
+                customOptions,
+                groupBy,
+                metric,
+                nodeType,
+              }) => (
+                <React.Fragment>
+                  <EuiFlexItem grow={false}>
+                    <WaffleMetricControls
+                      metric={metric}
+                      options={options}
+                      onChange={changeMetric}
+                    />
+                  </EuiFlexItem>
+                  <EuiFlexItem grow={false}>
+                    <WaffleGroupByControls
+                      options={groupByOptions}
+                      groupBy={groupBy}
+                      nodeType={nodeType}
+                      onChange={changeGroupBy}
+                      fields={createDerivedIndexPattern('metrics').fields}
+                      onChangeCustomOptions={changeCustomOptions}
+                      customOptions={customOptions}
+                    />
+                  </EuiFlexItem>
+                </React.Fragment>
+              )}
+            </WithWaffleOptions>
+          )}
+        </WithSource>
```

Comment:
For this overall component structure:

- WithSource
  - WithWaffleOptions
    - WaffleMetricControls
      - WaffleGroupByControls

It looks like the same structure / components are used across all of the different toolbars, just with different props. Could we consolidate this in to a shared component?

Reply:
@Kerry350, Chris mentioned something similar, what do you think about this solution? https://github.com/elastic/kibana/pull/47083/files#r332259476